### PR TITLE
Upgrade frontend to React 18

### LINKED ISSUE #961: Upgrade to React 18+
## Overview

We're currently on React 17. We'll eventually need to be on a newer major version, whether it's 18, or possibly higher.

### See

- https://reactjs.org/blog/2022/03/29/react-v18.html
- https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html

### PR DESCRIPTION
**Pull request by ammario**
Closes https://github.com/coder/coder/issues/961

### CODE REVIEW

**Comment by Kira-Pilot** [comment]
@ammario thanks for doing this! We wanted to discuss some of the React 18 changes in FE Variety before merging up - there are some rendering differences between 17 and 18. Do you mind if we hold off until then? We can mark it as a draft for now.

File: site/package.json
Comment (Kira-Pilot) [inline site/package.json:46]: Do you mind pinning these?

File: site/src/__mocks__/react-markdown.tsx
Comment (Kira-Pilot) [inline site/src/__mocks__/react-markdown.tsx:3]: Can we import `PropsWithChildren` directly from React?

Comment (Kira-Pilot) [reply site/src/__mocks__/react-markdown.tsx:3]: Ah, maybe you used a codemod. 
I assume we can also: 
````
const Component: FC<Props> = ({children}) => {...}
````
but maybe that wasn't an option?

**Comment by ammario** [comment]
> @ammario thanks for doing this! We wanted to discuss some of the React 18 changes in FE Variety before merging up - there are some rendering differences between 17 and 18. Do you mind if we hold off until then? We can mark it as a draft for now.

How about I fix these rendering concerns right now?

File: site/src/__mocks__/react-markdown.tsx
Comment (ammario) [reply site/src/__mocks__/react-markdown.tsx]: @Kira-Pilot `children` has to be explicit in the props now.


Also, I can import `PropsWithChildren` directly but it would be tedious for me to visit each file. I did use a codemod for the first pass.

If required for merge I will suck that up and make all the changes.

Comment (Kira-Pilot) [reply site/src/__mocks__/react-markdown.tsx]: No, not required - just curious

Comment (ammario) [reply site/src/__mocks__/react-markdown.tsx]: Ok great :)

**Comment by Kira-Pilot** [comment]
> How about I fix these rendering concerns right now?

Well, it's not so much fixing them (although there might be some of that) so much as it is understanding them. Here is the [React article](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-strict-mode) discussing the changes we are most curious about. Some questions I have (although I haven't finished reading this through) are:
 - how will this work when running tests locally, especially considering we often make use of Jest's `toBeCalledTimes`?
 - Are there workarounds? Should we be limiting use of `useEffect` (probably)? There's a whole issue thread [here](https://github.com/facebook/react/issues/24502).

None of that necessarily needs to be addressed in this PR - I just think usually we give folks a chance to read through the change logs before upgrading.
Would you like to join our next FE variety? We all committed to discuss the reading then. It would be nice to hear your perspective considering you did the upgrade already!

**Comment by presleyp** [comment]
Yeah we had decided to discuss this in FE V so I'd like to go ahead and do that. It's best to take tickets that have been groomed - this one was still in Enqueue because we were waiting on the discussion.

**Comment by ammario** [comment]
Ok. I'm happy to attend the next one. Please add me to the invite.

**Comment by ammario** [comment]
I intend on handling this end to end so it shouldn't add any work to your plates.

**Comment by Kira-Pilot** [comment]
@presleyp can add you to the invite!

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePageView.tsx
Comment (Kira-Pilot) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePageView.tsx:100]: Instead of rendering a fragment here if the condition isn't met, we should be able to do this:
````
       {!!props.createWorkspaceErrors[CreateWorkspaceErrors.GET_TEMPLATES_ERROR] && (
          <ErrorSummary
            error={props.createWorkspaceErrors[CreateWorkspaceErrors.GET_TEMPLATES_ERROR]}
          />
        )}
````

The error message you're encountering is TS's confusing way of telling you that it's not evaluating the condition as a boolean, so it needs you to explicitly cast it as such.

Comment (ammario) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePageView.tsx:100]: Gotcha. I thought the ternary was clearer, but the `!!` syntax makes sense too.

**Review by Kira-Pilot — APPROVED** [review]

**Comment by Kira-Pilot** [comment]
@ammario this is good to go when you're ready to merge up. The `act` warnings work has been captured in https://github.com/coder/coder/issues/3626.

**Comment by ammario** [comment]
Thank you @Kira-Pilot. You deserve the clout on the git history for this one.

**Pull request merged**
